### DISCUSSION TITLE
Hints for Chapter 8 Exercise: Rich Buttons

### DISCUSSION THREAD

**Discussion by Spacetoaster** [browserengineering/book#676] [Q&A] [answered]
I can't seem to figure out how to solve the "Rich Buttons" exercise of chapter 8 and was hoping someone can provide me with some pointers. From reading the chapter i know that i have to recurse on the button's children, but i can't figure out how to leverage the existing layouts in order to solve the exercise.
My naive idea was to have another InlineLayout which lays out the button's children and add that as a child in InputLayout, but i couldn't get it to work so far and feel like i'm on the wrong track.

**Comment by pavpanchekha** [comment, marked as answer]
The intended solution was what you did—have an `InlineLayout` inside the button to lay out its children. What problem were you running into? One tricky / important thing to get right is the order in which the `InputLayout` computes its fields and recurses into the child layouts.

**Reply by Spacetoaster** [reply]
Thanks, this was already enough to point me in the right direction.
Previously i was getting a bit frustrated with edge-cases like LineLayout expecting a `font` property during layout, which doesn't exist anymore on a button, if it has it's own InlineLayout (at least in my implementation). It just didn't feel like the right approach at the time and i got a bit frustrated, but knowing how it's intended gave me the motivation to get it working. Thanks again! Definitely one of the trickier exercises for me (so far 😄)

**Reply by pavpanchekha** [reply]
Definitely a trickier exercise. Generally speaking the exercises go from easier to harder, and there's a reason this one is the last one for this chapter :)
